Lib/inspect.py

```diff
@@ -2118,6 +2122,16 @@ def p(name_node, default_node, default=empty):
 
     return cls(parameters, return_annotation=cls.empty)
 
+def _get_type_hints(func):
+    try:
+        return typing.get_type_hints(func)
+    except Exception:
+        # First, try to use the get_type_hints to resolve
+        # annotations. But for keeping the behavior intact
+        # if there was a problem with that (like the namespace
+        # can't resolve some annotation) continue to use
+        # string annotations
+        return func.__annotations__
```

Comment:
I worry that that exception is going to mask errors (and e.g. cause us to update tests from `a: int` to `a: 'int'`).

I wonder if it would be better to just return `{}` in this case (given that we don't want to bubble up the error)?

Also, catching `Exception` feels wrong -- but I do agree that there are lots of different errors you could get (I managed to trigger TypeError, AttributeError, NameError, SyntaxError, IndentationError, IndentationError -- so who knows what else it could raise. (Just mumbling to myself here. :-)

Reply:
Actually, there is not much we can do for the exceptions since there is a code path that is going to `eval`.
```
(.venv) (Python 3.10.0a0) [ 12:31ÖÖ ]  [ isidentical@desktop:~/cpython/cpython(bpo-38605✗) ]
 $ python t.py  
(a: 'b()')
(.venv) (Python 3.10.0a0) [ 12:31ÖÖ ]  [ isidentical@desktop:~/cpython/cpython(bpo-38605✗) ]
 $ ./python t.py
Traceback (most recent call last):
  File "/home/isidentical/cpython/cpython/t.py", line 4, in <module>
    print(inspect.signature(x))
  File "/home/isidentical/cpython/cpython/Lib/inspect.py", line 3133, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
  File "/home/isidentical/cpython/cpython/Lib/inspect.py", line 2882, in from_callable
    return _signature_from_callable(obj, sigcls=cls,
  File "/home/isidentical/cpython/cpython/Lib/inspect.py", line 2333, in _signature_from_callable
    return _signature_from_function(sigcls, obj,
  File "/home/isidentical/cpython/cpython/Lib/inspect.py", line 2178, in _signature_from_function
    annotations = _get_type_hints(func)
  File "/home/isidentical/cpython/cpython/Lib/inspect.py", line 2127, in _get_type_hints
    return typing.get_type_hints(func)
  File "/home/isidentical/cpython/cpython/Lib/typing.py", line 1420, in get_type_hints
    value = _eval_type(value, globalns, localns)
  File "/home/isidentical/cpython/cpython/Lib/typing.py", line 255, in _eval_type
    return t._evaluate(globalns, localns, recursive_guard)
  File "/home/isidentical/cpython/cpython/Lib/typing.py", line 501, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
  File "/home/isidentical/cpython/cpython/t.py", line 1, in b
    def b(): 1/0 
ZeroDivisionError: division by zero
(.venv) (Python 3.10.0a0) [ 12:31ÖÖ ]  [ isidentical@desktop:~/cpython/cpython(bpo-38605✗) ]
 $ cat t.py
def b(): 1/0 
def x(a: 'b()'): pass
import inspect
print(inspect.signature(x))
```

Reply:
But I still think maybe we should return `{}` if it raises.

Reply:
This is a bit problematic when we are in a different context than the module, like this;
```py
import inspect, typing

def foo():
    class F: ...
    def foo(bar: F): ...
    print(inspect.signature(foo))

foo()
```
This would be normally resolved when we call `get_type_hints` with `locals()` but we can't do it (at least not without changing the API / or doing some frame-level dark magic). (that is from a test case of ours).

Reply:
(AFK) Does inspect.signature() have a way to pass a namespace?

Reply:
> Does inspect.signature() have a way to pass a namespace?

Apparently not

Reply:
Okay, then let’s keep this as you have it now. What other open issues do we have in this PR before I approve and merge?

Reply:
I kept What's new entry for later, I can do that I guess?

Reply:
Yes please!

Reply:
> Yes please!

Okay, will send it in a ten minutes!

Reply:
Please mention that get_type_hints() and signature() will behave differently, at least.

Reply:
Okay, it should be up, would you mind checking it out @gvanrossum?